worlds/sc2/rules.py

```diff
@@ -29,6 +35,121 @@ def is_item_placement(self, state):
+
+    def world_has_protoss_ground_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_protoss_ground_unit
+
+    def world_has_protoss_air_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_protoss_air_unit
+
+    def weapon_armor_upgrade_count(self, upgrade_item: str, state: CollectionState) -> int:
+        assert upgrade_item in upgrade_bundle_inverted_lookup.keys()
+        count: int = 0
+        if self.generic_upgrade_missions > 0:
+            if not self.is_item_placement(state):
+                # Item pool filtering, W/A upgrades aren't items
+                return WEAPON_ARMOR_UPGRADE_MAX_LEVEL
+            else:
+                count += floor(
+                    (100 / self.generic_upgrade_missions)
+                    * (state.count_group("Missions", self.player) / self.total_mission_count())
+                )
+        count += state.count(upgrade_item, self.player)
+        count += state.count_from_list(upgrade_bundle_inverted_lookup[upgrade_item], self.player)
+        if upgrade_item == item_names.PROGRESSIVE_PROTOSS_SHIELDS:
+            count += max(
+                state.count(item_names.PROGRESSIVE_PROTOSS_GROUND_UPGRADE, self.player),
+                state.count(item_names.PROGRESSIVE_PROTOSS_AIR_UPGRADE, self.player)
+            )
+        if (
+                upgrade_item in item_groups.protoss_generic_upgrades
+                and state.has(item_names.QUATRO, self.player)
+        ):
+            count += 1
+        return count
+
+    def terran_army_weapon_armor_upgrade_count(self, state: CollectionState) -> int:
+        """
+        Minimum W/A upgrade level for unit classes present in the world
+        :param state:
+        :return:
+        """
+        count: int = WEAPON_ARMOR_UPGRADE_MAX_LEVEL
+        if self.world_has_barracks_unit():
+            count = min(
+                count,
+                self.weapon_armor_upgrade_count(item_names.PROGRESSIVE_TERRAN_INFANTRY_WEAPON, state),
+                self.weapon_armor_upgrade_count(item_names.PROGRESSIVE_TERRAN_INFANTRY_ARMOR, state)
+            )
+        if self.world_has_factory_unit():
+            count = min(
+                count,
+                self.weapon_armor_upgrade_count(item_names.PROGRESSIVE_TERRAN_VEHICLE_WEAPON, state),
+                self.weapon_armor_upgrade_count(item_names.PROGRESSIVE_TERRAN_VEHICLE_ARMOR, state)
+            )
+        if self.world_has_starport_unit():
+            count = min(
+                count,
+                self.weapon_armor_upgrade_count(item_names.PROGRESSIVE_TERRAN_SHIP_WEAPON, state),
+                self.weapon_armor_upgrade_count(item_names.PROGRESSIVE_TERRAN_SHIP_ARMOR, state)
+            )
+        return count
+
+    def terran_very_hard_mission_weapon_armor_upgrades(self, state: CollectionState) -> bool:
```

Comment:
Recommend eliminating this class of function and replacing uses with `self.X_min_weapon_armour_upgrade() >= self.very_hard_weapon_armour_level()` to cut down on jumping around / repeated patterns when reading logic.